src/dos/program_ls.cpp

```diff
@@ -22,14 +22,240 @@
 
 #include <string>
 
+#include "checks.h"
+#include "program_more_output.h"
 #include "shell.h"
 #include "string_utils.h"
+#include "../ints/int10.h"
+
+CHECK_NARROWING();
 
 void LS::Run()
 {
-	std::string tmp = "";
-	cmd->GetStringRemain(tmp);
-	char args[CMD_MAXLINE];
-	safe_strcpy(args, tmp.c_str());
-	first_shell->CMD_LS(args);
+	// Handle command line
+
+	if (HelpRequested()) {
+		MoreOutputStrings output(*this);
+		output.AddString(MSG_Get("PROGRAM_LS_HELP_LONG"));
+		output.Display();
+		return;
+	}
+
+	constexpr bool remove_if_found = true;
+	const bool has_option_all = cmd->FindExist("/a", remove_if_found);
+
+	std::vector<std::string> patterns = {};
+	cmd->FillVector(patterns);
+
+	// Make sure no other switches are supplied
+
+	std::string tmp_str = {};
+	if (cmd->FindStringBegin("/", tmp_str)) {
+		tmp_str = std::string("/") + tmp_str;
+		WriteOut(MSG_Get("SHELL_ILLEGAL_SWITCH"), tmp_str.c_str());
+		return;
+	}
+
+	// Prepare search parameters
+
+	FatAttributeFlags search_attr = FatAttributeFlags::NotVolume;
+	if (!has_option_all) {
+		search_attr.system = false;
+		search_attr.hidden = false;
+	}
+
+	if (patterns.empty()) {
+		patterns.push_back({});
+	}
+	for (auto& pattern : patterns) {
+		pattern = to_search_pattern(pattern.c_str());
+	}
+
+	// Search for files/directories according to pattern
+
+	std::vector<DOS_DTA::Result> dir_contents = {};
+
+	const RealPt original_dta = dos.dta();
+	dos.dta(dos.tables.tempdta);
+	DOS_DTA dta(dos.dta());
+
+	for (const auto& pattern : patterns) {
+		if (shutdown_requested) {
+			break;
+		}
+
+		if (!DOS_FindFirst(pattern.c_str(), search_attr)) {
+			continue;
+		}
+
+		do {
+			DOS_DTA::Result result = {};
+			dta.GetResult(result);
+			if (result.IsDummyDirectory()) {
+				continue;
+			}
+			dir_contents.push_back(result);
+		} while (!shutdown_requested && DOS_FindNext());
+	}
+
+	dos.dta(original_dta);
+	if (shutdown_requested) {
+		return;
+	}
+
+	if (dir_contents.empty()) {
+		WriteOut(MSG_Get("SHELL_NO_FILES_SUBDIRS_TO_DISPLAY"));
+		return;
+	}
+
+	// Sort directory contents
+
+	constexpr bool reverse_order = false;
+	DOS_Sort(dir_contents,
+	         ResultSorting::ByName,
+	         reverse_order,
+	         ResultGrouping::NonFilesFirst);
```

Comment:
Actually, I intend to rethink this API one day, so that we can support ordering in `DIR` command like in latest Windows: https://learn.microsoft.com/en-us/windows-server/administration/windows-commands/dir